src/main/java/reposense/report/ReportGenerator.java

```diff
@@ -119,59 +125,80 @@
      * Clone, analyze and generate the report for repositories in {@code repoLocationMap}.
      * Performs analysis and report generation of each repository in parallel with the cloning of the next repository.
      */
-    private static void cloneAndAnalyzeRepos(
-            Map<RepoLocation, List<RepoConfiguration>> repoLocationMap, String outputPath) throws IOException {
+    private static void cloneAndAnalyzeRepos(List<RepoConfiguration> configs, String outputPath) {
+        Map<RepoLocation, List<RepoConfiguration>> repoLocationMap = groupConfigsByRepoLocation(configs);
         RepoCloner repoCloner = new RepoCloner();
         RepoLocation clonedRepoLocation = null;
 
-        for (RepoLocation location : repoLocationMap.keySet()) {
-            repoCloner.clone(outputPath, repoLocationMap.get(location).get(0));
+        List<RepoLocation> repoLocationList = new ArrayList<>(repoLocationMap.keySet());
+        Set<RepoLocation> successfullyClonedRepos = new HashSet<>();
+
+        // clones another repo in parallel while analyzing a cloned repo (if any)
+        for (RepoLocation location : repoLocationList) {
```

Comment:
@eugenepeh do we need to worry about unsafe removal of `RepoConfiguration` elements here even as `RepoLocation` won't be removed?

Reply:
Nope, that is not necessary. As long as you can perform the removal without utilizing global variable, its fine.